### PR TITLE
Initial embedded ELF module loader.

### LINKED ISSUE #4717: LLVM codegen is pulling in libc/libm (malloc, free, floorf) functions and shouldn't be
Here's where the flag is set:

https://github.com/google/iree/blob/af39f98403a30b85442130f25ceec5eb53e0c60d/iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp#L127-L129

We punted on this back with https://github.com/google/iree/issues/2810 and now the flag is even leaking into our [Python documentation](https://github.com/google/iree/blob/main/bindings/python/pyiree/jax/README.md#ahead-of-time-compilation).

### LINKED ISSUE #5351: Implement simple cross-platform/arch ELF loader for executables
Today we are using platform-native shared libraries for dynamically loading compiled executables. This is a good path for tooling as then platform-native debuggers, profilers, and backtraces all work correctly. It's not a good path for deployment, though, as it requires us to embed a compiled artifact per architecture crossed with each target platform. Instead we should just use ELF everywhere and load ourselves. Our shared libraries are by design extremely simple and we control all of the linker parameters for them meaning that we only need to implement a tiny subset of features. There are still cases where we may want platform-native libraries (like iOS to satisfy code signing) but they would become opt-in with the generic ELF used for everything by default.

https://github.com/google/iree/issues/3845#issuecomment-727539414 mentions a few tricks. Relatively simple technique, just needs per-arch relocation shims and ABI thunks to marshal registers, which is about 100 lines of code/arch total and easy to incrementally add.

### PR DESCRIPTION
**Pull request by benvanik**
Enabled with `-iree-llvm-link-embedded` and `-iree-llvm-target-triple={any}-pc-linux-elf`.
This is still considered experimental and not enabled by default. More verification is needed across platforms and we need to make progress on #4717 before more models can run.

~1000x faster than the system loader on Windows (60-100ms -> 50us) and 64b + ELF memory usage. Most of the time spent on Windows/x64 is in applying page protections:
image

Though this is a massive performance win the real benefit is dramatically simplified compiler configuration and cross-platform retargetability: there's no need for the Android NDK/etc, cross-compilation (Linux->Windows->Mac->etc) is trivial, and the ELF files we generate will run anywhere C will run and we can PROT_EXEC. Users just need lld (ld may work, but eh) with the appropriate target architectures built into it and we will only need to embed one binary per architecture (vs one per architecture crossed with all platforms it may be used on). This is how we get portable AOT execution on the cheap.

Imports are not supported so it fails on any executable that ends up using -lm (floorf, etc).
The `iree-translate` invocation I'm using to produce the executables attempts to work around this in some cases with `fmaf`:
```
    -iree-llvm-target-triple=x86_64-pc-linux-elf \
    -iree-codegen-linalg-to-llvm-use-unfused-fma \
    -iree-llvm-link-embedded=true \
    -iree-llvm-debug-symbols=false
```

Tentative support is in place for bare-metal targets via the new `IREE_PLATFORM_GENERIC` selector; it's untested but likely to be just a short iteration from working. This path requires no syscalls (file IO, dlopen, mmap, etc) and just C11's aligned_alloc/free.

Fixes #5351.

### CODE REVIEW

File: iree/hal/local/elf/arch/riscv.c [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        break;
+      default:
+        return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                                "unimplemented riscv32 relocation type %08X",
+                                type);
+    }
+  }
+  return iree_ok_status();
+}
+#else
```
Comment (stellaraccident) [inline iree/hal/local/elf/arch/riscv.c:82]: Add a comment: RISCV64

(I had to scroll up to find what we are branching on)

File: iree/hal/local/elf/arch/x86_64.c [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/base/target_platform.h"
+#include "iree/hal/local/elf/arch.h"
+
+#if defined(IREE_ARCH_X86_64)
+
```
Comment (stellaraccident) [inline iree/hal/local/elf/arch/x86_64.c:19]: Is there a one stop document to reference here like the others?

File: iree/hal/local/elf/arch/x86_64.c [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  IREE_ELF_R_X86_64_PC64 = 24,      // Place relative 64-bit signed
+};
+
+static iree_status_t iree_elf_arch_x86_64_apply_rela(
+    iree_elf_relocation_state_t* state, iree_host_size_t rela_count,
+    const iree_elf_rela_t* rela_table) {
+  for (iree_host_size_t i = 0; i < rela_count; ++i) {
+    const iree_elf_rela_t* rela = &rela_table[i];
+    uint32_t type = IREE_ELF_R_TYPE(rela->r_info);
+    if (type == 0) continue;
```
Comment (stellaraccident) [inline iree/hal/local/elf/arch/x86_64.c:58]: Can you spell the '0' symbolically to make it more clear that you are early exiting for the same case covered below by the switch?

File: iree/hal/local/elf/arch/x86_64.c [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    iree_elf_addr_t sym_addr = 0;
+    if (IREE_ELF_R_SYM(rela->r_info) != 0) {
+      return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                              "symbol-relative relocations not implemented");
+    }
+
+    iree_elf_addr_t instr_ptr =
+        (iree_elf_addr_t)state->vaddr_bias + rela->r_offset;
+    switch (type) {
+      case IREE_ELF_R_X86_64_NONE:
```
Comment (stellaraccident) [inline iree/hal/local/elf/arch/x86_64.c:70]: This case is unreachable. (I'm ok keeping it for clarity but note)

File: iree/hal/local/elf/arch/x86_64_msvc.asm [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  push rbp
+  mov rbp, rsp
+  push rbx
+  push rdi
+  push rsi
+  push r12
+  push r13
+  push r14
+  push r15
+
```
Comment (stellaraccident) [inline iree/hal/local/elf/arch/x86_64_msvc.asm:44]: Add a TODO for the XMM regs?

Comment (stellaraccident) [reply iree/hal/local/elf/arch/x86_64_msvc.asm:44]: (here and below)

File: iree/hal/local/elf/elf_module.c [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+        .offset = phdr->p_vaddr,
+        .length = phdr->p_memsz,
+    };
+    IREE_RETURN_IF_ERROR(iree_memory_view_commit_ranges(
+        module->vaddr_bias, 1, &byte_range,
+        IREE_MEMORY_ACCESS_READ | IREE_MEMORY_ACCESS_WRITE));
+
+    // Copy data present in the file.
+    if (phdr->p_filesz > 0) {
+      memcpy(module->vaddr_bias + phdr->p_vaddr, raw_data.data + phdr->p_offset,
```
Comment (stellaraccident) [inline iree/hal/local/elf/elf_module.c:271]: Comment that as a future optimization, we can overlay the memory from fixed mmaps (on archs where this is convenient and an actual optimization)?

File: iree/hal/local/elf/platform/generic.c [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+iree_status_t iree_memory_view_protect_ranges(void* base_address,
+                                              iree_host_size_t range_count,
+                                              const iree_byte_range_t* ranges,
+                                              iree_memory_access_t new_access) {
+  // No-op.
+  return iree_ok_status();
+}
+
+void iree_memory_view_flush_icache(void* base_address,
+                                   iree_host_size_t length) {
```
Comment (stellaraccident) [inline iree/hal/local/elf/platform/generic.c:95]: Maybe undef CLEAR_CACHE if already defined by someone else?

File: iree/hal/local/elf/platform/linux.c [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+      break;
+    }
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+void iree_memory_view_flush_icache(void* base_address,
+                                   iree_host_size_t length) {
```
Comment (stellaraccident) [inline iree/hal/local/elf/platform/linux.c:145]: Ditto on undefing CLEAR_CACHE

File: iree/hal/local/loaders/embedded_library_loader.c [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        IREE_STATUS_FAILED_PRECONDITION,
+        "executable does not support this version of the runtime (%d)",
+        IREE_HAL_EXECUTABLE_LIBRARY_LATEST_VERSION);
+  }
+  const iree_hal_executable_library_header_t* header =
+      *executable->library.header;
+
+  // Ensure that if the library is built for a particular sanitizer that we also
+  // were compiled with that sanitizer enabled.
+  switch (header->sanitizer) {
```
Comment (stellaraccident) [inline iree/hal/local/loaders/embedded_library_loader.c:67]: It seems like ultimately, whether the executable was compiled for a sanitizer should be a multi-versioning flag that we select an executable based on (i.e. same as detecting a variant based on cpu features).

**Review by stellaraccident — APPROVED** [review]
This is really awesome (and a good walk down memory lane for me). Thanks for the side chat with clarifications on specifics.

File: iree/hal/local/elf/elf_types.h [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+typedef uint16_t iree_elf32_half_t;
+typedef uint32_t iree_elf32_off_t;
+typedef int32_t iree_elf32_sword_t;
+typedef uint32_t iree_elf32_word_t;
+
+typedef uint8_t iree_elf64_byte_t;
+typedef uint64_t iree_elf64_addr_t;
+typedef uint16_t iree_elf64_half_t;
+typedef uint64_t iree_elf64_off_t;
+typedef int32_t iree_elf64_sword_t;
```
Comment (ScottTodd) [inline iree/hal/local/elf/elf_types.h:60]: ⚔️  `sword`  :O

**Review by ScottTodd — APPROVED** [review]
I just scanned through, did not review all files in detail. Looks good :D

File: iree/hal/local/elf/platform/generic.c (hunk 7 above)
Comment (benvanik) [reply iree/hal/local/elf/platform/generic.c:95]: Changed this to `IREE_ELF_CLEAR_CACHE` and now explicitly allowing it to be defined externally (should make it easier for people to target weirder platforms without code changes).

File: iree/hal/local/elf/arch/x86_64_msvc.asm (hunk 5 above)
Comment (benvanik) [reply iree/hal/local/elf/arch/x86_64_msvc.asm:44]: Decided just to do it here - done!

File: iree/hal/local/loaders/embedded_library_loader.c (hunk 9 above)
Comment (benvanik) [reply iree/hal/local/loaders/embedded_library_loader.c:67]: Changed this to always fail if any sanitizers were compiled in - getting compilerrt stable across platforms is .... not worth it :P

**Pull request merged**